silkrpc/json/types.cpp

```diff
@@ -305,20 +306,36 @@ void to_json(nlohmann::json& json, const Receipt& receipt) {
 void from_json(const nlohmann::json& json, Receipt& receipt) {
     SILKRPC_TRACE << "from_json<Receipt> json: " << json.dump() << "\n";
     if (json.is_array()) {
-        if (json.size() < 3) {
+        int start_index = 0;
+        if (json.size() == 4) {
+            if (!json[start_index].is_number()) {
+                auto message = str(boost::format("Receipt CBOR: number expected in [%d]") % start_index);
+                throw std::system_error{std::make_error_code(std::errc::invalid_argument), ""};
+            }
+            receipt.type = json[start_index];
+            start_index++;
+        } else if (json.size() != 3) {
             throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Receipt CBOR: missing entries"};
         }
-        if (!json[0].is_null()) {
-            throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Receipt CBOR: null expected in [0]"};
+
+        if (!json[start_index].is_null()) {
+            auto message = str(boost::format("Receipt CBOR: null expected in [%d]") % start_index);
+            throw std::system_error{std::make_error_code(std::errc::invalid_argument), message};
         }
-        if (!json[1].is_number()) {
-            throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Receipt CBOR: number expected in [1]"};
+
+        start_index++;
+        if (!json[start_index].is_number()) {
+            auto message = str(boost::format("Receipt CBOR: number expected in [%d]") % start_index);
+            throw std::system_error{std::make_error_code(std::errc::invalid_argument), message};
         }
-        receipt.success = json[1] == 1u;
-        if (!json[2].is_number()) {
-            throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Receipt CBOR: number expected in [2]"};
+        receipt.success = json[start_index] == 1u;
+
+        start_index++;
+        if (!json[start_index].is_number()) {
+            auto message = str(boost::format("Receipt CBOR: number expected in [%d]") % start_index);
+            throw std::system_error{std::make_error_code(std::errc::invalid_argument), message};
         }
-        receipt.cumulative_gas_used = json[2];
+        receipt.cumulative_gas_used = json[start_index];
```

Comment:
I would go with the hard-coded solution handling 4 CBOR fields because 1) we really don't need backward compatibility with old format 2) readability is worse this way IMHO 3) we avoid introducing dependency on boost::format